File: frontend/src/api/system.ts
import client from "./client";

export interface ServerConfig {
  id: number;
  interface_name: string;
  public_key: string;
  listen_port: number;
  address: string;
  dns: string | null;
  mtu: number | null;
  post_up: string | null;
  post_down: string | null;
  endpoint: string;
  client_retry_enabled: boolean;
  admin_port: number;
  api_http_port: number;
  api_https_port: number;
  api_http_enabled: boolean;
  vpn_domain: string | null;
}

export interface ServerConfigUpdate {
  address?: string;
  listen_port?: number;
  endpoint?: string;
  dns?: string | null;
  mtu?: number | null;
  post_up?: string | null;
  post_down?: string | null;
  client_retry_enabled?: boolean;
  admin_port?: number;
  api_http_port?: number;
  api_https_port?: number;
  api_http_enabled?: boolean;
  vpn_domain?: string | null;
}

export interface RegenerateKeyResponse {
  public_key: string;
  message: string;
}

export interface HealthResponse {
  status: string;
  db: string;
  wg_interface: string;
  tunnel_count: number;
  uptime_seconds: number;
  is_initialized: boolean;
}

export interface SystemMetrics {
  ram_percent: number;
  ram_used_gb: number;
  ram_total_gb: number;
  cpu_percent: number;
  cpu_count: number;
  disk_percent: number;
  disk_used_gb: number;
  disk_total_gb: number;
}

export interface ADConfig {
  ad_enabled: boolean;
  ad_server: string | null;
  ad_server_backup: string | null;
  ad_base_dn: string | null;
  ad_bind_dn: string | null;
  ad_user_filter: string | null;
  ad_group_filter: string | null;
  ad_use_ssl: boolean;
  ad_require_membership: boolean;
}

export interface ADConfigUpdate {
  ad_enabled?: boolean;
  ad_server?: string | null;
  ad_server_backup?: string | null;
  ad_base_dn?: string | null;
  ad_bind_dn?: string | null;
  ad_bind_password?: string | null;
  ad_user_filter?: string | null;
  ad_group_filter?: string | null;
  ad_use_ssl?: boolean;
  ad_require_membership?: boolean;
}

export interface ADGroupMapping {
  id: number;
  ad_group_dn: string;
  ad_group_name: string;
  netloom_group_id: number;
  netloom_group_name: string | null;
  enabled: boolean;
  priority: number;
}

export interface ADGroupMappingCreate {
  ad_group_dn: string;
  ad_group_name: string;
  netloom_group_id: number;
  enabled?: boolean;
  priority?: number;
}

export interface ADGroupFromAD {
  dn: string;
  name: string;
  sam_name: string;
}

export const systemApi = {
  health: () => client.get<HealthResponse>("/system/health"),
  getServerConfig: () => client.get<ServerConfig>("/system/server-config"),
  updateServerConfig: (data: ServerConfigUpdate) =>
    client.put<ServerConfig>("/system/server-config", data),
  regenerateKey: () => client.post<RegenerateKeyResponse>("/system/server-config/regenerate-key"),
  applyConfig: () => client.post<{ message: string }>("/system/apply-config"),
  wgUp: () => client.post<{ message: string }>("/system/wg/up"),
  wgDown: () => client.post<{ message: string }>("/system/wg/down"),
  wgRestart: () => client.post<{ message: string }>("/system/wg/restart"),
  backup: () => client.post<{ message: string; path: string }>("/system/backup"),
  getMetrics: () => client.get<SystemMetrics>("/system/metrics"),
  exportBackup: async () => {
    const response = await client.get("/system/backup/export", { responseType: "blob" });
    const contentDisposition = response.headers["content-disposition"] ?? "";
    const match = contentDisposition.match(/filename="([^"]+)"/);
    const filename = match ? match[1] : "netloom_backup.zip";
    const url = URL.createObjectURL(response.data as Blob);
    const a = document.createElement("a");
    a.href = url;
    a.download = filename;
    a.click();
    URL.revokeObjectURL(url);
  },
  importBackup: (file: File) => {
    const formData = new FormData();
    formData.append("file", file);
    return client.post("/system/backup/import", formData, {
      headers: { "Content-Type": "multipart/form-data" },
    });
  },
  getADConfig: () => client.get<ADConfig>("/system/ad"),
  updateADConfig: (data: ADConfigUpdate) => client.put<ADConfig>("/system/ad", data),
  testADConnection: () => client.post<{ success: boolean; error?: string; message?: string }>("/system/ad/test-connection"),
  getADGroupsFromServer: () => client.get<{ groups: ADGroupFromAD[] }>("/system/ad/groups"),
  getADGroupMappings: () => client.get<ADGroupMapping[]>("/system/ad/group-mappings"),
  createADGroupMapping: (data: ADGroupMappingCreate) => client.post<ADGroupMapping>("/system/ad/group-mappings", data),
  bulkCreateADGroupMappings: (mappings: ADGroupMappingCreate[]) => 
    client.post<ADGroupMapping[]>("/system/ad/group-mappings/bulk", { mappings }),
  updateADGroupMapping: (id: number, data: Partial<ADGroupMapping>) => 
    client.patch<ADGroupMapping>(`/system/ad/group-mappings/${id}`, data),
  deleteADGroupMapping: (id: number) => client.delete(`/system/ad/group-mappings/${id}`),
  resetOnboarding: () => client.post<{ message: string }>("/onboarding/reset"),
};

File: backend/app/schemas/system.py
from pydantic import BaseModel


class ServerConfigResponse(BaseModel):
    id: int
    interface_name: str
    public_key: str
    listen_port: int
    address: str
    dns: str | None
    mtu: int | None
    post_up: str | None
    post_down: str | None
    endpoint: str
    client_retry_enabled: bool = False
    admin_port: int = 7777
    api_http_port: int = 7771
    api_https_port: int = 7772
    api_http_enabled: bool = True
    vpn_domain: str | None = None

    model_config = {"from_attributes": True}


class ServerConfigUpdate(BaseModel):
    address: str | None = None
    listen_port: int | None = None
    dns: str | None = None
    mtu: int | None = None
    post_up: str | None = None
    post_down: str | None = None
    endpoint: str | None = None
    client_retry_enabled: bool | None = None
    admin_port: int | None = None
    api_http_port: int | None = None
    api_https_port: int | None = None
    api_http_enabled: bool | None = None
    vpn_domain: str | None = None


class RegenerateKeyResponse(BaseModel):
    public_key: str
    message: str


class HealthResponse(BaseModel):
    status: str
    db: str
    wg_interface: str
    tunnel_count: int
    uptime_seconds: int
    is_initialized: bool = True

    model_config = {"from_attributes": True}


class BackupResponse(BaseModel):
    message: str
    path: str


class FirewallStatusResponse(BaseModel):
    firewall_enabled: bool


class FirewallStatusUpdate(BaseModel):
    firewall_enabled: bool


class SystemMetricsResponse(BaseModel):
    ram_percent: float
    ram_used_gb: float
    ram_total_gb: float
    cpu_percent: float
    cpu_count: int
    disk_percent: float
    disk_used_gb: float
    disk_total_gb: float


class TLSConfigResponse(BaseModel):
    tls_enabled: bool
    tls_cert_path: str
    tls_key_path: str
    tls_host: str
    tls_port: int
    http_port: int
    tls_auto_generate: bool
    tls_cert_days: int
    tls_country: str
    tls_state: str
    tls_locality: str
    tls_organization: str
    tls_common_name: str
    cert_exists: bool
    key_exists: bool

    model_config = {"from_attributes": False}


class TLSConfigUpdate(BaseModel):
    tls_enabled: bool | None = None
    tls_cert_path: str | None = None
    tls_key_path: str | None = None
    tls_host: str | None = None
    tls_port: int | None = None
    http_port: int | None = None
    tls_auto_generate: bool | None = None
    tls_cert_days: int | None = None
    tls_country: str | None = None
    tls_state: str | None = None
    tls_locality: str | None = None
    tls_organization: str | None = None
    tls_common_name: str | None = None


class TLSRegenerateCertResponse(BaseModel):
    message: str
    cert_path: str
    key_path: str


class ADConfigResponse(BaseModel):
    ad_enabled: bool
    ad_server: str | None
    ad_server_backup: str | None
    ad_base_dn: str | None
    ad_bind_dn: str | None
    ad_user_filter: str | None
    ad_group_filter: str | None
    ad_use_ssl: bool
    ad_require_membership: bool

    model_config = {"from_attributes": True}


class ADConfigUpdate(BaseModel):
    ad_enabled: bool | None = None
    ad_server: str | None = None
    ad_server_backup: str | None = None
    ad_base_dn: str | None = None
    ad_bind_dn: str | None = None
    ad_bind_password: str | None = None
    ad_user_filter: str | None = None
    ad_group_filter: str | None = None
    ad_use_ssl: bool | None = None
    ad_require_membership: bool | None = None


class ADGroupMappingResponse(BaseModel):
    id: int
    ad_group_dn: str
    ad_group_name: str
    netloom_group_id: int
    netloom_group_name: str | None = None
    enabled: bool
    priority: int

    model_config = {"from_attributes": True}


class ADGroupMappingCreate(BaseModel):
    ad_group_dn: str
    ad_group_name: str
    netloom_group_id: int
    enabled: bool = True
    priority: int = 0


class ADGroupMappingUpdate(BaseModel):
    netloom_group_id: int | None = None
    enabled: bool | None = None
    priority: int | None = None


class ADGroupMappingBulkCreate(BaseModel):
    mappings: list[ADGroupMappingCreate]


class ADGroupsFromADResponse(BaseModel):
    groups: list[dict]
